### PR TITLE
Use miette in dlint for diagnostic output

### PR DESCRIPTION
**Pull request by bartlomieju**
Closes https://github.com/denoland/deno_lint/issues/1024

Screenshot_2022-03-16_at_23 52 57

Co-Authored-By: 강동윤 <kdy1997.dev@gmail.com>

### CODE REVIEW

**Comment by bartlomieju** [comment]
@kdy1 you did a heavy lifting in SWC to integrate miette and I reused some of the code. I'll co-author you if this PR lands.

File: Cargo.lock [hunk 1]
```diff
@@ -133,6 +153,12 @@ version = "3.8.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "8f1e260c3a9040a7c19a12468758f4c16f31a81a1fe087482be9570ec864bb6c"
 
+[[package]]
+name = "cc"
```
Comment (ry) [inline Cargo.lock:157]: hmm..

**Comment by kdy1** [comment]
@bartlomieju Thank you!

File: Cargo.lock (hunk 1 above)
Comment (kitsonk) [reply Cargo.lock:157]: hmm indeed... native C code is a necessary dependency? 😕

Comment (kitsonk) [reply Cargo.lock:157]: Look like it comes from [backtrace](https://crates.io/crates/backtrace), surprising given that backtrace is an "official" crate...

Comment (kitsonk) [reply Cargo.lock:157]: Looks like it is only used when targetting android, there is a single .c binding file.  There is this note in the [Cargo.toml](https://github.com/rust-lang/backtrace-rs/blob/master/Cargo.toml#L53-L56):

```toml
[build-dependencies]
# Only needed for Android, but cannot be target dependent
# https://github.com/rust-lang/cargo/issues/4932
cc = "1.0.67"
```

Comment (bartlomieju) [reply Cargo.lock:157]: Opened https://github.com/zkat/miette/issues/132

**Comment by bartlomieju** [comment]
This is now ready, just pending new release of `miette`

**Review by dsherret — COMMENTED** [review]
Almost there. It seems there is one bug.

File: examples/dlint/diagnostics.rs [hunk 2]
```diff
@@ -1,200 +1,169 @@
...
+    let src_end = self
+      .source
+      .line_end(end_line_column.line_index + context_lines_after)
+      .0 as usize;
+    let src_end = std::cmp::min(src_end, line_count);
+    let src = &self.source.text_str()[src_start..src_end];
+    let name = Some(self.filename.to_string());
+    let start = miette::SourceOffset::from(src_start);
+    let len = miette::SourceOffset::from(src_end - src_start);
+    let span = miette::SourceSpan::new(start, len);
```
[the remark this answers is not in the document]
Comment (bartlomieju) [reply examples/dlint/diagnostics.rs:123]: Haven't found any, but I also didn't look very hard - this is mostly copy-pasted from SWC source code, so I trusted that kdy1 found the shortest way to do it.

[the remark this answers is not in the document]
Comment (kdy1) [reply examples/dlint/diagnostics.rs:123]: We can extract it as a method, but I didn't bother to just because there it was not required frequently.

**Comment by bartlomieju** [comment]
Looks like the tests fail because the output on CI is different than the output I get locally (additional info like links are printed). I need to tweak some more settings in `miette` to enforce consistent behavior.

**Comment by bartlomieju** [comment]
@dsherret please take a look

**Review by dsherret — APPROVED** [review]
LGTM

**Pull request merged**